README.md
<!-- MANPAGE: BEGIN EXCLUDED SECTION -->
<div align="center">

[![pragy-dl](https://raw.githubusercontent.com/pragy-dl/pragy-dl/master/.github/banner.svg)](#readme)

[![Release version](https://img.shields.io/github/v/release/pragy-dl/pragy-dl?color=brightgreen&label=Download&style=for-the-badge)](#installation "Installation")
[![PyPi](https://img.shields.io/badge/-PyPi-blue.svg?logo=pypi&labelColor=555555&style=for-the-badge)](https://pypi.org/project/pragy-dl "PyPi")
[![Donate](https://img.shields.io/badge/_-Donate-red.svg?logo=githubsponsors&labelColor=555555&style=for-the-badge)](Collaborators.md#collaborators "Donate")
[![Matrix](https://img.shields.io/matrix/pragy-dl:matrix.org?color=brightgreen&labelColor=555555&label=&logo=element&style=for-the-badge)](https://matrix.to/#/#pragy-dl:matrix.org "Matrix")
[![Discord](https://img.shields.io/discord/807245652072857610?color=blue&labelColor=555555&label=&logo=discord&style=for-the-badge)](https://discord.gg/H5MNcFW63r "Discord")
[![Supported Sites](https://img.shields.io/badge/-Supported_Sites-brightgreen.svg?style=for-the-badge)](supportedsites.md "Supported Sites")
[![License: Unlicense](https://img.shields.io/badge/-Unlicense-blue.svg?style=for-the-badge)](LICENSE "License")
[![CI Status](https://img.shields.io/github/actions/workflow/status/pragy-dl/pragy-dl/core.yml?branch=master&label=Tests&style=for-the-badge)](https://github.com/pragy-dl/pragy-dl/actions "CI Status")
[![Commits](https://img.shields.io/github/commit-activity/m/pragy-dl/pragy-dl?label=commits&style=for-the-badge)](https://github.com/pragy-dl/pragy-dl/commits "Commit History")
[![Last Commit](https://img.shields.io/github/last-commit/pragy-dl/pragy-dl/master?label=&style=for-the-badge&display_timestamp=committer)](https://github.com/pragy-dl/pragy-dl/pulse/monthly "Last activity")

</div>
<!-- MANPAGE: END EXCLUDED SECTION -->

pragy-dl is a [youtube-dl](https://github.com/ytdl-org/youtube-dl) fork based on the now inactive [youtube-dlc](https://github.com/blackjack4494/yt-dlc). The main focus of this project is adding new features and patches while also keeping up to date with the original project

<!-- MANPAGE: MOVE "USAGE AND OPTIONS" SECTION HERE -->

<!-- MANPAGE: BEGIN EXCLUDED SECTION -->
* [NEW FEATURES](#new-features)
    * [Differences in default behavior](#differences-in-default-behavior)
* [INSTALLATION](#installation)
    * [Detailed instructions](https://github.com/pragy-dl/pragy-dl/wiki/Installation)
    * [Update](#update)
    * [Release Files](#release-files)
    * [Dependencies](#dependencies)
    * [Compile](#compile)
* [USAGE AND OPTIONS](#usage-and-options)
    * [General Options](#general-options)
    * [Network Options](#network-options)
    * [Geo-restriction](#geo-restriction)
    * [Video Selection](#video-selection)
    * [Download Options](#download-options)
    * [Filesystem Options](#filesystem-options)
    * [Thumbnail Options](#thumbnail-options)
    * [Internet Shortcut Options](#internet-shortcut-options)
    * [Verbosity and Simulation Options](#verbosity-and-simulation-options)
    * [Workarounds](#workarounds)
    * [Video Format Options](#video-format-options)
    * [Subtitle Options](#subtitle-options)
    * [Authentication Options](#authentication-options)
    * [Post-processing Options](#post-processing-options)
    * [SponsorBlock Options](#sponsorblock-options)
    * [Extractor Options](#extractor-options)
* [CONFIGURATION](#configuration)
    * [Configuration file encoding](#configuration-file-encoding)
    * [Authentication with netrc](#authentication-with-netrc)
    * [Notes about environment variables](#notes-about-environment-variables)
* [OUTPUT TEMPLATE](#output-template)
    * [Output template examples](#output-template-examples)
* [FORMAT SELECTION](#format-selection)
    * [Filtering Formats](#filtering-formats)
    * [Sorting Formats](#sorting-formats)
    * [Format Selection examples](#format-selection-examples)
* [MODIFYING METADATA](#modifying-metadata)
    * [Modifying metadata examples](#modifying-metadata-examples)
* [EXTRACTOR ARGUMENTS](#extractor-arguments)
* [PLUGINS](#plugins)
    * [Installing Plugins](#installing-plugins)
    * [Developing Plugins](#developing-plugins)
* [EMBEDDING pragy-dl](#embedding-pragy-dl)
    * [Embedding examples](#embedding-examples)
* [DEPRECATED OPTIONS](#deprecated-options)
* [CONTRIBUTING](CONTRIBUTING.md#contributing-to-pragy-dl)
    * [Opening an Issue](CONTRIBUTING.md#opening-an-issue)
    * [Developer Instructions](CONTRIBUTING.md#developer-instructions)
* [WIKI](https://github.com/pragy-dl/pragy-dl/wiki)
    * [FAQ](https://github.com/pragy-dl/pragy-dl/wiki/FAQ)
<!-- MANPAGE: END EXCLUDED SECTION -->


# NEW FEATURES

* Forked from [**yt-dlc@f9401f2**](https://github.com/blackjack4494/yt-dlc/commit/f9401f2a91987068139c5f757b12fc711d4c0cee) and merged with [**youtube-dl@66ab08**](https://github.com/ytdl-org/youtube-dl/commit/66ab0814c4baa2dc79c2dd5287bc0ad61a37c5b9) ([exceptions](https://github.com/pragy-dl/pragy-dl/issues/21))

* **[SponsorBlock Integration](#sponsorblock-options)**: You can mark/remove sponsor sections in YouTube videos by utilizing the [SponsorBlock](https://sponsor.ajay.app) API

* **[Format Sorting](#sorting-formats)**: The default format sorting options have been changed so that higher resolution and better codecs will be now preferred instead of simply using larger bitrate. Furthermore, you can now specify the sort order using `-S`. This allows for much easier format selection than what is possible by simply using `--format` ([examples](#format-selection-examples))

* **Merged with animelover1984/youtube-dl**: You get most of the features and improvements from [animelover1984/youtube-dl](https://github.com/animelover1984/youtube-dl) including `--write-comments`, `BiliBiliSearch`, `BilibiliChannel`, Embedding thumbnail in mp4/ogg/opus, playlist infojson etc. Note that NicoNico livestreams are not available. See [#31](https://github.com/pragy-dl/pragy-dl/pull/31) for details.

* **YouTube improvements**:
    * Supports Clips, Stories (`ytstories:<channel UCID>`), Search (including filters)**\***, YouTube Music Search, Channel-specific search, Search prefixes (`ytsearch:`, `ytsearchdate:`)**\***, Mixes, and Feeds (`:ytfav`, `:ytwatchlater`, `:ytsubs`, `:ythistory`, `:ytrec`, `:ytnotif`)
    * Fix for [n-sig based throttling](https://github.com/ytdl-org/youtube-dl/issues/29326) **\***
    * Supports some (but not all) age-gated content without cookies
    * Download livestreams from the start using `--live-from-start` (*experimental*)
    * `255kbps` audio is extracted (if available) from YouTube Music when premium cookies are given
    * Channel URLs download all uploads of the channel, including shorts and live

* **Cookies from browser**: Cookies can be automatically extracted from all major web browsers using `--cookies-from-browser BROWSER[+KEYRING][:PROFILE][::CONTAINER]`

* **Download time range**: Videos can be downloaded partially based on either timestamps or chapters using `--download-sections`

* **Split video by chapters**: Videos can be split into multiple files based on chapters using `--split-chapters`

* **Multi-threaded fragment downloads**: Download multiple fragments of m3u8/mpd videos in parallel. Use `--concurrent-fragments` (`-N`) option to set the number of threads used

* **Aria2c with HLS/DASH**: You can use `aria2c` as the external downloader for DASH(mpd) and HLS(m3u8) formats

* **New and fixed extractors**: Many new extractors have been added and a lot of existing ones have been fixed. See the [changelog](Changelog.md) or the [list of supported sites](supportedsites.md)

* **New MSOs**: Philo, Spectrum, SlingTV, Cablevision, RCN etc.

* **Subtitle extraction from manifests**: Subtitles can be extracted from streaming media manifests. See [commit/be6202f](https://github.com/pragy-dl/pragy-dl/commit/be6202f12b97858b9d716e608394b51065d0419f) for details

* **Multiple paths and output templates**: You can give different [output templates](#output-template) and download paths for different types of files. You can also set a temporary path where intermediary files are downloaded to using `--paths` (`-P`)

* **Portable Configuration**: Configuration files are automatically loaded from the home and root directories. See [CONFIGURATION](#configuration) for details

* **Output template improvements**: Output templates can now have date-time formatting, numeric offsets, object traversal etc. See [output template](#output-template) for details. Even more advanced operations can also be done with the help of `--parse-metadata` and `--replace-in-metadata`

* **Other new options**: Many new options have been added such as `--alias`, `--print`, `--concat-playlist`, `--wait-for-video`, `--retry-sleep`, `--sleep-requests`, `--convert-thumbnails`, `--force-download-archive`, `--force-overwrites`, `--break-match-filter` etc

* **Improvements**: Regex and other operators in `--format`/`--match-filter`, multiple `--postprocessor-args` and `--downloader-args`, faster archive checking, more [format selection options](#format-selection), merge multi-video/audio, multiple `--config-locations`, `--exec` at different stages, etc

* **Plugins**: Extractors and PostProcessors can be loaded from an external file. See [plugins](#plugins) for details

* **Self updater**: The releases can be updated using `pragy-dl -U`, and downgraded using `--update-to` if required

* **Nightly builds**: [Automated nightly builds](#update-channels) can be used with `--update-to nightly`

See [changelog](Changelog.md) or [commits](https://github.com/pragy-dl/pragy-dl/commits) for the full list of changes

Features marked with a **\*** have been back-ported to youtube-dl

### Differences in default behavior

Some of pragy-dl's default options are different from that of youtube-dl and youtube-dlc:

* pragy-dl supports only [Python 3.7+](## "Windows 7"), and *may* remove support for more versions as they [become EOL](https://devguide.python.org/versions/#python-release-cycle); while [youtube-dl still supports Python 2.6+ and 3.2+](https://github.com/ytdl-org/youtube-dl/issues/30568#issue-1118238743)
* The options `--auto-number` (`-A`), `--title` (`-t`) and `--literal` (`-l`), no longer work. See [removed options](#Removed) for details
* `avconv` is not supported as an alternative to `ffmpeg`
* pragy-dl stores config files in slightly different locations to youtube-dl. See [CONFIGURATION](#configuration) for a list of correct locations
* The default [output template](#output-template) is `%(title)s [%(id)s].%(ext)s`. There is no real reason for this change. This was changed before pragy-dl was ever made public and now there are no plans to change it back to `%(title)s-%(id)s.%(ext)s`. Instead, you may use `--compat-options filename`
* The default [format sorting](#sorting-formats) is different from youtube-dl and prefers higher resolution and better codecs rather than higher bitrates. You can use the `--format-sort` option to change this to any order you prefer, or use `--compat-options format-sort` to use youtube-dl's sorting order
* The default format selector is `bv*+ba/b`. This means that if a combined video + audio format that is better than the best video-only format is found, the former will be preferred. Use `-f bv+ba/b` or `--compat-options format-spec` to revert this
* Unlike youtube-dlc, pragy-dl does not allow merging multiple audio/video streams into one file by default (since this conflicts with the use of `-f bv*+ba`). If needed, this feature must be enabled using `--audio-multistreams` and `--video-multistreams`. You can also use `--compat-options multistreams` to enable both
* `--no-abort-on-error` is enabled by default. Use `--abort-on-error` or `--compat-options abort-on-error` to abort on errors instead
* When writing metadata files such as thumbnails, description or infojson, the same information (if available) is also written for playlists. Use `--no-write-playlist-metafiles` or `--compat-options no-playlist-metafiles` to not write these files

Changelog.md
# Changelog

<!--
# To create a release, dispatch the https://github.com/pragy-dl/pragy-dl/actions/workflows/release.yml workflow on master
-->

### 2023.10.07

#### Extractor changes
- **abc.net.au**: iview: [Improve `episode` extraction](https://github.com/pragy-dl/pragy-dl/commit/a9efb4b8d74f3583450ffda0ee57259a47d39c70) ([#8201](https://github.com/pragy-dl/pragy-dl/issues/8201)) by [xofe](https://github.com/xofe)
- **erocast**: [Add extractor](https://github.com/pragy-dl/pragy-dl/commit/47c598783c98c179e04dd12c2a3fee0f3dc53087) ([#8264](https://github.com/pragy-dl/pragy-dl/issues/8264)) by [madewokherd](https://github.com/madewokherd)
- **gofile**: [Fix token cookie bug](https://github.com/pragy-dl/pragy-dl/commit/0730d5a966fa8a937d84bfb7f68be5198acb039b) by [bashonly](https://github.com/bashonly)
- **iq.com**: [Fix extraction and subtitles](https://github.com/pragy-dl/pragy-dl/commit/35d9cbaf9638ccc9daf8a863063b2e7c135bc664) ([#8260](https://github.com/pragy-dl/pragy-dl/issues/8260)) by [AS6939](https://github.com/AS6939)
- **lbry**
    - [Add playlist support](https://github.com/pragy-dl/pragy-dl/commit/48cceec1ddb8649b5e771df8df79eb9c39c82b90) ([#8213](https://github.com/pragy-dl/pragy-dl/issues/8213)) by [bashonly](https://github.com/bashonly), [drzraf](https://github.com/drzraf), [Grub4K](https://github.com/Grub4K)
    - [Extract `uploader_id`](https://github.com/pragy-dl/pragy-dl/commit/0e722f2f3ca42e634fd7b06ee70b16bf833ce132) ([#8244](https://github.com/pragy-dl/pragy-dl/issues/8244)) by [drzraf](https://github.com/drzraf)
- **litv**: [Fix extractor](https://github.com/pragy-dl/pragy-dl/commit/91a670a4f7babe9c8aa2018f57d8c8952a6f49d8) ([#7785](https://github.com/pragy-dl/pragy-dl/issues/7785)) by [jiru](https://github.com/jiru)
- **neteasemusic**: [Fix extractors](https://github.com/pragy-dl/pragy-dl/commit/f980df734cf5c0eaded2f7b38c6c60bccfeebb48) ([#8181](https://github.com/pragy-dl/pragy-dl/issues/8181)) by [c-basalt](https://github.com/c-basalt)
- **nhk**: [Fix VOD extraction](https://github.com/pragy-dl/pragy-dl/commit/e831c80e8b2fc025b3b67d82974cc59e3526fdc8) ([#8249](https://github.com/pragy-dl/pragy-dl/issues/8249)) by [garret1317](https://github.com/garret1317)
- **radiko**: [Improve extraction](https://github.com/pragy-dl/pragy-dl/commit/2ad3873f0dfa9285c91d2160e36c039e69d597c7) ([#8221](https://github.com/pragy-dl/pragy-dl/issues/8221)) by [garret1317](https://github.com/garret1317)
- **substack**
    - [Fix download cookies bug](https://github.com/pragy-dl/pragy-dl/commit/2f2dda3a7e85148773da3cdbc03ac9949ec1bc45) ([#8219](https://github.com/pragy-dl/pragy-dl/issues/8219)) by [handlerug](https://github.com/handlerug)
    - [Fix embed extraction](https://github.com/pragy-dl/pragy-dl/commit/fbcc299bd8a19cf8b3c8805d6c268a9110230973) ([#8218](https://github.com/pragy-dl/pragy-dl/issues/8218)) by [handlerug](https://github.com/handlerug)
- **theta**: [Remove extractors](https://github.com/pragy-dl/pragy-dl/commit/792f1e64f6a2beac51e85408d142b3118115c4fd) ([#8251](https://github.com/pragy-dl/pragy-dl/issues/8251)) by [alerikaisattera](https://github.com/alerikaisattera)
- **wrestleuniversevod**: [Call API with device ID](https://github.com/pragy-dl/pragy-dl/commit/b095fd3fa9d58a65dc9b830bd63b9d909422aa86) ([#8272](https://github.com/pragy-dl/pragy-dl/issues/8272)) by [bashonly](https://github.com/bashonly)
- **xhamster**: user: [Support creator urls](https://github.com/pragy-dl/pragy-dl/commit/cc8d8441524ec3442d7c0d3f8f33f15b66aa06f3) ([#8232](https://github.com/pragy-dl/pragy-dl/issues/8232)) by [Grub4K](https://github.com/Grub4K)
- **youtube**
    - [Fix `heatmap` extraction](https://github.com/pragy-dl/pragy-dl/commit/03e85ea99db76a2fddb65bf46f8819bda780aaf3) ([#8299](https://github.com/pragy-dl/pragy-dl/issues/8299)) by [bashonly](https://github.com/bashonly)
    - [Raise a warning for `Incomplete Data` instead of an error](https://github.com/pragy-dl/pragy-dl/commit/eb5bdbfa70126c7d5355cc0954b63720522e462c) ([#8238](https://github.com/pragy-dl/pragy-dl/issues/8238)) by [coletdjnz](https://github.com/coletdjnz)

#### Misc. changes
- **cleanup**
    - [Update extractor tests](https://github.com/pragy-dl/pragy-dl/commit/19c90e405b4137c06dfe6f9aaa02396df0da93e5) ([#7718](https://github.com/pragy-dl/pragy-dl/issues/7718)) by [trainman261](https://github.com/trainman261)
    - Miscellaneous: [377e85a](https://github.com/pragy-dl/pragy-dl/commit/377e85a1797db9e98b78b38203ed9d4ded229991) by [dirkf](https://github.com/dirkf), [gamer191](https://github.com/gamer191), [Grub4K](https://github.com/Grub4K)

### 2023.09.24

#### Important changes
- **The minimum *recommended* Python version has been raised to 3.8**
Since Python 3.7 has reached end-of-life, support for it will be dropped soon. [Read more](https://github.com/pragy-dl/pragy-dl/issues/7803)
- Security: [[CVE-2023-40581](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2023-40581)] [Prevent RCE when using `--exec` with `%q` on Windows](https://github.com/pragy-dl/pragy-dl/security/advisories/GHSA-42h4-v29r-42qg)
    - The shell escape function is now using `""` instead of `\"`.
    - `utils.Popen` has been patched to properly quote commands.

#### Core changes
- [Fix HTTP headers and cookie handling](https://github.com/pragy-dl/pragy-dl/commit/6c5211cebeacfc53ad5d5ddf4a659be76039656f) by [bashonly](https://github.com/bashonly), [pukkandan](https://github.com/pukkandan)
- [Fix `--check-formats`](https://github.com/pragy-dl/pragy-dl/commit/8cb7fc44db010e965d808ee679ef0725cb6e147c) by [pukkandan](https://github.com/pukkandan)
- [Fix support for upcoming Python 3.12](https://github.com/pragy-dl/pragy-dl/commit/836e06d246512f286f30c1371b2c54b72c9ecd93) ([#8130](https://github.com/pragy-dl/pragy-dl/issues/8130)) by [Grub4K](https://github.com/Grub4K)
- [Merged with youtube-dl 66ab08](https://github.com/pragy-dl/pragy-dl/commit/9d6254069c75877bc88bc3584f4326fb1853a543) by [coletdjnz](https://github.com/coletdjnz)

pragy_dl/update.py
import atexit
import contextlib
import hashlib
import json
import os
import platform
import re
import subprocess
import sys
from zipimport import zipimporter

from .compat import functools  # isort: split
from .compat import compat_realpath, compat_shlex_quote
from .networking import Request
from .networking.exceptions import HTTPError, network_exceptions
from .utils import (
    Popen,
    cached_method,
    deprecation_warning,
    remove_end,
    remove_start,
    shell_quote,
    system_identifier,
    version_tuple,
)
from .version import CHANNEL, UPDATE_HINT, VARIANT, __version__

UPDATE_SOURCES = {
    'stable': 'pragy-dl/pragy-dl',
    'nightly': 'pragy-dl/pragy-dl-nightly-builds',
}
REPOSITORY = UPDATE_SOURCES['stable']

_VERSION_RE = re.compile(r'(\d+\.)*\d+')

API_BASE_URL = 'https://api.github.com/repos'

# Backwards compatibility variables for the current channel
API_URL = f'{API_BASE_URL}/{REPOSITORY}/releases'


@functools.cache
def _get_variant_and_executable_path():
    """@returns (variant, executable_path)"""
    if getattr(sys, 'frozen', False):
        path = sys.executable
        if not hasattr(sys, '_MEIPASS'):
            return 'py2exe', path
        elif sys._MEIPASS == os.path.dirname(path):
            return f'{sys.platform}_dir', path
        elif sys.platform == 'darwin':
            machine = '_legacy' if version_tuple(platform.mac_ver()[0]) < (10, 15) else ''
        else:
            machine = f'_{platform.machine().lower()}'
            # Ref: https://en.wikipedia.org/wiki/Uname#Examples
            if machine[1:] in ('x86', 'x86_64', 'amd64', 'i386', 'i686'):
                machine = '_x86' if platform.architecture()[0][:2] == '32' else ''
        return f'{remove_end(sys.platform, "32")}{machine}_exe', path

    path = os.path.dirname(__file__)
    if isinstance(__loader__, zipimporter):
        return 'zip', os.path.join(path, '..')
    elif (os.path.basename(sys.argv[0]) in ('__main__.py', '-m')
          and os.path.exists(os.path.join(path, '../.git/HEAD'))):
        return 'source', path
    return 'unknown', path


def detect_variant():
    return VARIANT or _get_variant_and_executable_path()[0]


@functools.cache
def current_git_head():
    if detect_variant() != 'source':
        return
    with contextlib.suppress(Exception):
        stdout, _, _ = Popen.run(
            ['git', 'rev-parse', '--short', 'HEAD'],
            text=True, cwd=os.path.dirname(os.path.abspath(__file__)),
            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        if re.fullmatch('[0-9a-f]+', stdout.strip()):
            return stdout.strip()


_FILE_SUFFIXES = {
    'zip': '',
    'py2exe': '_min.exe',
    'win_exe': '.exe',
    'win_x86_exe': '_x86.exe',
    'darwin_exe': '_macos',
    'darwin_legacy_exe': '_macos_legacy',
    'linux_exe': '_linux',
    'linux_aarch64_exe': '_linux_aarch64',
    'linux_armv7l_exe': '_linux_armv7l',
}

_NON_UPDATEABLE_REASONS = {
    **{variant: None for variant in _FILE_SUFFIXES},  # Updatable
    **{variant: f'Auto-update is not supported for unpackaged {name} executable; Re-download the latest release'
       for variant, name in {'win32_dir': 'Windows', 'darwin_dir': 'MacOS', 'linux_dir': 'Linux'}.items()},
    'source': 'You cannot update when running from source code; Use git to pull the latest changes',
    'unknown': 'You installed pragy-dl with a package manager or setup.py; Use that to update',
    'other': 'You are using an unofficial build of pragy-dl; Build the executable again',
}


def is_non_updateable():
    if UPDATE_HINT:
        return UPDATE_HINT
    return _NON_UPDATEABLE_REASONS.get(
        detect_variant(), _NON_UPDATEABLE_REASONS['unknown' if VARIANT else 'other'])


def _get_system_deprecation():
    MIN_SUPPORTED, MIN_RECOMMENDED = (3, 7), (3, 8)

    if sys.version_info > MIN_RECOMMENDED:
        return None

    major, minor = sys.version_info[:2]
    if sys.version_info < MIN_SUPPORTED:
        msg = f'Python version {major}.{minor} is no longer supported'
    else:
        msg = f'Support for Python version {major}.{minor} has been deprecated. '
        # Temporary until `win_x86_exe` uses 3.8, which will deprecate Vista and Server 2008
        if detect_variant() == 'win_x86_exe':
            platform_name = platform.platform()
            if any(platform_name.startswith(f'Windows-{name}') for name in ('Vista', '2008Server')):
                msg = 'Support for Windows Vista/Server 2008 has been deprecated. '
            else:
                return None
        msg += ('See  https://github.com/pragy-dl/pragy-dl/issues/7803  for details.'
                '\nYou may stop receiving updates on this version at any time')

    major, minor = MIN_RECOMMENDED
    return f'{msg}! Please update to Python {major}.{minor} or above'


def _sha256_file(path):
    h = hashlib.sha256()
    mv = memoryview(bytearray(128 * 1024))
    with open(os.path.realpath(path), 'rb', buffering=0) as f:
        for n in iter(lambda: f.readinto(mv), 0):
            h.update(mv[:n])
    return h.hexdigest()


class Updater:
    _exact = True

    def __init__(self, ydl, target=None):
        self.ydl = ydl

        self.target_channel, sep, self.target_tag = (target or CHANNEL).rpartition('@')
        # stable => stable@latest
        if not sep and ('/' in self.target_tag or self.target_tag in UPDATE_SOURCES):
            self.target_channel = self.target_tag
            self.target_tag = None
        elif not self.target_channel:
            self.target_channel = CHANNEL.partition('@')[0]

        if not self.target_tag:
            self.target_tag = 'latest'
            self._exact = False
        elif self.target_tag != 'latest':
            self.target_tag = f'tags/{self.target_tag}'

        if '/' in self.target_channel:
            self._target_repo = self.target_channel
            if self.target_channel not in (CHANNEL, *UPDATE_SOURCES.values()):
                self.ydl.report_warning(
                    f'You are switching to an {self.ydl._format_err("unofficial", "red")} executable '
                    f'from {self.ydl._format_err(self._target_repo, self.ydl.Styles.EMPHASIS)}. '
                    f'Run {self.ydl._format_err("at your own risk", "light red")}')
                self._block_restart('Automatically restarting into custom builds is disabled for security reasons')
        else:
            self._target_repo = UPDATE_SOURCES.get(self.target_channel)
            if not self._target_repo:
                self._report_error(
                    f'Invalid update channel {self.target_channel!r} requested. '
                    f'Valid channels are {", ".join(UPDATE_SOURCES)}', True)

    def _version_compare(self, a, b, channel=CHANNEL):
        if self._exact and channel != self.target_channel:
            return False

        if _VERSION_RE.fullmatch(f'{a}.{b}'):
            a, b = version_tuple(a), version_tuple(b)
            return a == b if self._exact else a >= b
        return a == b

    @functools.cached_property
    def _tag(self):
        if self._version_compare(self.current_version, self.latest_version):
            return self.target_tag

        identifier = f'{detect_variant()} {self.target_channel} {system_identifier()}'
        for line in self._download('_update_spec', 'latest').decode().splitlines():
            if not line.startswith('lock '):
                continue
            _, tag, pattern = line.split(' ', 2)
            if re.match(pattern, identifier):
                if not self._exact:
                    return f'tags/{tag}'
                elif self.target_tag == 'latest' or not self._version_compare(
                        tag, self.target_tag[5:], channel=self.target_channel):
                    self._report_error(
                        f'pragy-dl cannot be updated above {tag} since you are on an older Python version', True)
                    return f'tags/{self.current_version}'
        return self.target_tag

    @cached_method
    def _get_version_info(self, tag):
        url = f'{API_BASE_URL}/{self._target_repo}/releases/{tag}'
        self.ydl.write_debug(f'Fetching release info: {url}')
        return json.loads(self.ydl.urlopen(Request(url, headers={
            'Accept': 'application/vnd.github+json',
            'User-Agent': 'pragy-dl',
            'X-GitHub-Api-Version': '2022-11-28',
        })).read().decode())

    @property
    def current_version(self):
        """Current version"""
        return __version__

    @staticmethod
    def _label(channel, tag):
        """Label for a given channel and tag"""
        return f'{channel}@{remove_start(tag, "tags/")}'

    def _get_actual_tag(self, tag):
        if tag.startswith('tags/'):
            return tag[5:]
        return self._get_version_info(tag)['tag_name']

    @property
    def new_version(self):
        """Version of the latest release we can update to"""
        return self._get_actual_tag(self._tag)

    @property
    def latest_version(self):
        """Version of the target release"""
        return self._get_actual_tag(self.target_tag)

    @property
    def has_update(self):
        """Whether there is an update available"""
        return not self._version_compare(self.current_version, self.new_version)

    @functools.cached_property
    def filename(self):
        """Filename of the executable"""
        return compat_realpath(_get_variant_and_executable_path()[1])

    def _download(self, name, tag):
        slug = 'latest/download' if tag == 'latest' else f'download/{tag[5:]}'
        url = f'https://github.com/{self._target_repo}/releases/{slug}/{name}'
        self.ydl.write_debug(f'Downloading {name} from {url}')
        return self.ydl.urlopen(url).read()

    @functools.cached_property
    def release_name(self):
        """The release filename"""
        return f'pragy-dl{_FILE_SUFFIXES[detect_variant()]}'

    @functools.cached_property
    def release_hash(self):
        """Hash of the latest release"""
        hash_data = dict(ln.split()[::-1] for ln in self._download('SHA2-256SUMS', self._tag).decode().splitlines())
        return hash_data[self.release_name]

    def _report_error(self, msg, expected=False):
        self.ydl.report_error(msg, tb=False if expected else None)
        self.ydl._download_retcode = 100

    def _report_permission_error(self, file):
        self._report_error(f'Unable to write to {file}; Try running as administrator', True)

    def _report_network_error(self, action, delim=';'):
        self._report_error(
            f'Unable to {action}{delim} visit  '
            f'https://github.com/{self._target_repo}/releases/{self.target_tag.replace("tags/", "tag/")}', True)

    def check_update(self):
        """Report whether there is an update available"""
        if not self._target_repo:
            return False
        try:
            self.ydl.to_screen((
                f'Available version: {self._label(self.target_channel, self.latest_version)}, ' if self.target_tag == 'latest' else ''
            ) + f'Current version: {self._label(CHANNEL, self.current_version)}')
        except network_exceptions as e:
            return self._report_network_error(f'obtain version info ({e})', delim='; Please try again later or')

        if not is_non_updateable():
            self.ydl.to_screen(f'Current Build Hash: {_sha256_file(self.filename)}')

        if self.has_update:
            return True

        if self.target_tag == self._tag:
            self.ydl.to_screen(f'pragy-dl is up to date ({self._label(CHANNEL, self.current_version)})')
        elif not self._exact:
            self.ydl.report_warning('pragy-dl cannot be updated any further since you are on an older Python version')
        return False

    def update(self):
        """Update pragy-dl executable to the latest version"""
        if not self.check_update():
            return
        err = is_non_updateable()
        if err:
            return self._report_error(err, True)
        self.ydl.to_screen(f'Updating to {self._label(self.target_channel, self.new_version)} ...')
        if (_VERSION_RE.fullmatch(self.target_tag[5:])
                and version_tuple(self.target_tag[5:]) < (2023, 3, 2)):
            self.ydl.report_warning('You are downgrading to a version without --update-to')
            self._block_restart('Cannot automatically restart to a version without --update-to')

        directory = os.path.dirname(self.filename)
        if not os.access(self.filename, os.W_OK):
            return self._report_permission_error(self.filename)
        elif not os.access(directory, os.W_OK):
            return self._report_permission_error(directory)

        new_filename, old_filename = f'{self.filename}.new', f'{self.filename}.old'
        if detect_variant() == 'zip':  # Can be replaced in-place
            new_filename, old_filename = self.filename, None

        try:
            if os.path.exists(old_filename or ''):
                os.remove(old_filename)
        except OSError:
            return self._report_error('Unable to remove the old version')

        try:
            newcontent = self._download(self.release_name, self._tag)
        except network_exceptions as e:
            if isinstance(e, HTTPError) and e.status == 404:
                return self._report_error(
                    f'The requested tag {self._label(self.target_channel, self.target_tag)} does not exist', True)
            return self._report_network_error(f'fetch updates: {e}')

        try:
            expected_hash = self.release_hash
        except Exception:
            self.ydl.report_warning('no hash information found for the release')
        else:
            if hashlib.sha256(newcontent).hexdigest() != expected_hash:
                return self._report_network_error('verify the new executable')

        try:
            with open(new_filename, 'wb') as outf:
                outf.write(newcontent)
        except OSError:
            return self._report_permission_error(new_filename)

        if old_filename:
            mask = os.stat(self.filename).st_mode
            try:
                os.rename(self.filename, old_filename)
            except OSError:
                return self._report_error('Unable to move current version')

            try:
                os.rename(new_filename, self.filename)
            except OSError:
                self._report_error('Unable to overwrite current version')
                return os.rename(old_filename, self.filename)

        variant = detect_variant()
        if variant.startswith('win') or variant == 'py2exe':
            atexit.register(Popen, f'ping 127.0.0.1 -n 5 -w 1000 & del /F "{old_filename}"',
                            shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        elif old_filename:
            try:
                os.remove(old_filename)
            except OSError:
                self._report_error('Unable to remove the old version')

            try:
                os.chmod(self.filename, mask)
            except OSError:
                return self._report_error(
                    f'Unable to set permissions. Run: sudo chmod a+rx {compat_shlex_quote(self.filename)}')

        self.ydl.to_screen(f'Updated pragy-dl to {self._label(self.target_channel, self.new_version)}')
        return True

    @functools.cached_property
    def cmd(self):
        """The command-line to run the executable, if known"""
        # There is no sys.orig_argv in py < 3.10. Also, it can be [] when frozen
        if getattr(sys, 'orig_argv', None):
            return sys.orig_argv
        elif getattr(sys, 'frozen', False):
            return sys.argv

    def restart(self):
        """Restart the executable"""
        assert self.cmd, 'Must be frozen or Py >= 3.10'
        self.ydl.write_debug(f'Restarting: {shell_quote(self.cmd)}')
        _, _, returncode = Popen.run(self.cmd)
        return returncode

    def _block_restart(self, msg):
        def wrapper():
            self._report_error(f'{msg}. Restart pragy-dl to use the updated version', expected=True)
            return self.ydl._download_retcode
        self.restart = wrapper


def run_update(ydl):
    """Update the program file with the latest version from the repository
    @returns    Whether there was a successful update (No update = False)
    """
    return Updater(ydl).update()


# Deprecated
def update_self(to_screen, verbose, opener):
    import traceback

    deprecation_warning(f'"{__name__}.update_self" is deprecated and may be removed '
                        f'in a future version. Use "{__name__}.run_update(ydl)" instead')

    printfn = to_screen

    class FakeYDL():
        to_screen = printfn

        def report_warning(self, msg, *args, **kwargs):
            return printfn(f'WARNING: {msg}', *args, **kwargs)

        def report_error(self, msg, tb=None):
            printfn(f'ERROR: {msg}')
            if not verbose:
                return
            if tb is None:
                # Copied from YoutubeDL.trouble
                if sys.exc_info()[0]:
                    tb = ''
                    if hasattr(sys.exc_info()[1], 'exc_info') and sys.exc_info()[1].exc_info[0]:
                        tb += ''.join(traceback.format_exception(*sys.exc_info()[1].exc_info))
                    tb += traceback.format_exc()
                else:
                    tb_data = traceback.format_list(traceback.extract_stack())
                    tb = ''.join(tb_data)
            if tb:
                printfn(tb)

        def write_debug(self, msg, *args, **kwargs):
            printfn(f'[debug] {msg}', *args, **kwargs)

        def urlopen(self, url):
            return opener.open(url)

    return run_update(FakeYDL())


__all__ = ['Updater']
